pylutron_caseta/pairing.py

```diff
@@ -0,0 +1,192 @@
+)
+
+PAIR_KEY = "key"
+PAIR_CERT = "cert"
+PAIR_CA = "ca"
+PAIR_VERSION = "version"
+
+
+def _generate_private_key():
+    LOGGER.info("Generating a new private key...")
+    return rsa.generate_private_key(
+        public_exponent=65537, key_size=2048, backend=default_backend()
+    )
+
+
+def _convert_private_key_to_pem(private_key):
+    return private_key.private_bytes(
+        encoding=serialization.Encoding.PEM,
+        format=serialization.PrivateFormat.PKCS8,
+        encryption_algorithm=serialization.NoEncryption(),
+    )
+
+
+def _generate_csr(private_key):
+    return (
+        x509.CertificateSigningRequestBuilder()
+        .subject_name(CERT_SUBJECT)
+        .sign(private_key, hashes.SHA256(), default_backend())
+    )
+
+
+class JsonSocket:
+    """A socket that reads and writes json objects."""
+
+    def __init__(self, raw_socket):
+        """Create a JsonSocket wrapping the provided socket."""
+        self._socket = raw_socket
+
+    def read_json(self):
+        """Read an object."""
+        buffer = b""
+        while not buffer.endswith(b"\r\n"):
+            buffer += self._socket.read(1024)
+
+        LOGGER.debug("received: %s", buffer)
+        return json.loads(buffer.decode("UTF-8"))
+
+    def write_json(self, obj):
+        """Write an object."""
+        buffer = ("%s\r\n" % json.dumps(obj)).encode("ASCII")
+        self._socket.write(buffer)
+        LOGGER.debug("sent: %s", buffer)
+
+
+def _generate_certificate(server_addr, ssl_context, csr):
+    with socket.create_connection(
+        (server_addr, 8083), timeout=SOCKET_TIMEOUT
+    ) as raw_socket:
+        raw_socket.settimeout(BUTTON_PRESS_TIMEOUT)
+        with ssl_context.wrap_socket(raw_socket) as tls_socket:
+            json_socket = JsonSocket(tls_socket)
+
+            LOGGER.debug("Connected to bridge.")
+            LOGGER.info(
+                "Press the small black button on the back of the Caseta bridge..."
+            )
+            while True:
+                message = json_socket.read_json()
+                if message.get("Header", {}).get("ContentType", "").startswith(
+                    "status;"
+                ) and "PhysicalAccess" in (
+                    message.get("Body", {}).get("Status", {}).get("Permissions", [])
+                ):
+                    break
+
+            LOGGER.debug("Getting my certificate...")
+            csr_text = csr.public_bytes(serialization.Encoding.PEM).decode("ASCII")
+            json_socket.write_json(
+                {
+                    "Header": {
+                        "RequestType": "Execute",
+                        "Url": "/pair",
+                        "ClientTag": "get-cert",
+                    },
+                    "Body": {
+                        "CommandType": "CSR",
+                        "Parameters": {
+                            "CSR": csr_text,
+                            "DisplayName": "get_lutron_cert.py",
```

Comment:
I'm not sure what this DisplayName is for, but it should probably be pylutron-caseta or settable by the caller.

Reply:
Adjusted to `pylutron_caseta`